runners/samza/src/main/java/org/apache/beam/runners/samza/metrics/SamzaMetricsContainer.java

```diff
@@ -53,24 +59,43 @@ public class SamzaMetricsContainer {
 
   private static final Logger LOG = LoggerFactory.getLogger(SamzaMetricsContainer.class);
   private static final String BEAM_METRICS_GROUP = "BeamMetrics";
-  // global metrics container is the default container that can be used in user threads
   public static final String GLOBAL_CONTAINER_STEP_NAME = "GLOBAL_METRICS";
   public static final String USE_SHORT_METRIC_NAMES_CONFIG =
       "beam.samza.metrics.useShortMetricNames";
   public static final String COMMIT_ALL_METRIC_UPDATES =
       "beam.samza.metrics.commitAllMetricUpdates";
+  public static final String DEFER_TO_EXECUTOR_CONFIG = "beam.samza.metrics.deferToExecutor";
+  public static final String METRIC_UPDATE_INTERVAL_SEC_CONFIG =
+      "beam.samza.metrics.updateIntervalSec";
 
   private final MetricsContainerStepMap metricsContainers = new MetricsContainerStepMap();
   private final MetricsRegistryMap metricsRegistry;
   private final boolean useShortMetricNames;
   private final boolean commitAllMetricUpdates;
+  private final boolean deferToExecutor;
+  private final Set<String> activeStepNames = ConcurrentHashMap.newKeySet();
+  // technically this doesn't need to be volatile, but needed to pass spotbugs check
+  private volatile boolean executorServiceStarted = false;
+  private ScheduledExecutorService executorService;
+  private ScheduledFuture<?> scheduledFuture;
+  private final long metricUpdateIntervalSec;
 
   public SamzaMetricsContainer(MetricsRegistryMap metricsRegistry, Config config) {
```

Comment:
Are we sure this SamzaMetricsContainer is a process-wise singleton?  It might end up creating multiple thread pools if it's not. Any race conditions on updating metrics if it's not?

Reply:
I’m not entirely sure if SamzaMetricsContainer is a process-wide singleton. From what I see, it’s instantiated in SamzaExecutionContext, which should only occur once per container/JVM. However, I also see instances in both BoundedSourceSystem and UnboundedSourceSystem.

Having multiple thread pools might be fine—it could result in one per container plus the number of sources. If necessary, we can optimize this by moving it to a static instantiation. However, I think this can be addressed in a follow-up PR if needed.

As for race conditions, I don’t believe this code introduces any. If multiple instances of SamzaMetricsContainer exist, each with its own threads updating metrics, the original code would have been equally, if not more, susceptible to race conditions due to the frequency of updates. So, the current implementation shouldn’t increase that risk.

Reply:
Update see this comment: https://github.com/linkedin/beam/pull/128/files#r1772470875, doing this should prevent us from introducing additional race conditions.

Reply:
Right it seems like UnboundedSourceSystem will have a separate SamzaMetricsContainer for the input step, while SamzaExecutionContext have a SamzaMetricsContainer for the rest of the DoFns. Theoretically, there should be no overlap on the steps they try to update. There could be concurrent calls on the underlying MetricsRegistry.